### ISSUE TITLE
ModifierOrder: false negative on module-info.java files

### ISSUE THREAD

**Issue by vivek-0509** [checkstyle/checkstyle#21166]
Parent Issue #21113

I have read check documentation: https://checkstyle.org/checks/modifier/modifierorder.html
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

Note: module-info.java is only parseable after #8240, so this is reproduced on that branch.

```bash
vivek@Viveks-MacBook-Air checkstyle % cat > config.xml <<'EOF'
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
    "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
    "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
    <module name="ModifierOrder"/>
  </module>
</module>
EOF

vivek@Viveks-MacBook-Air checkstyle % cat > mo/module-info.java <<'EOF'
module com.example.app {
    requires static transitive java.sql;
}
EOF

vivek@Viveks-MacBook-Air checkstyle % RUN_LOCALE="-Duser.language=en -Duser.country=US"
java $RUN_LOCALE -jar target/checkstyle-13.10.0-SNAPSHOT-all.jar -c config.xml mo/module-info.java

Starting audit...
Audit done.
```

Out of order modifiers on a class member are reported:

```bash
vivek@Viveks-MacBook-Air checkstyle % cat > A.java <<'EOF'
package com.example;

class A {
    static public int x = 1;
}
EOF

vivek@Viveks-MacBook-Air checkstyle % java $RUN_LOCALE -jar target/checkstyle-13.10.0-SNAPSHOT-all.jar -c config.xml A.java

Starting audit...
[ERROR] A.java:4:12: 'public' modifier out of order with the JLS suggestions. [ModifierOrder]
Audit done.
Checkstyle ends with 1 errors.
```

**Describe what you expect in detail.**

Modifiers on a `requires` directive should be validated the same way modifiers on a class member are, so the module input should be reported.

The [Google Java Style Guide, module declaration section](https://google.github.io/styleguide/javaguide.html#s3.5-module-declaration) states:

> Modifiers on `requires` module directives, when present, appear in the following order: `transitive static`

The Check does not see them because its only token is `MODIFIERS`, while module modifiers are flat children of `REQUIRES`, since JLS 7.7 defines no ModuleModifier production:

```
REQUIRES -> requires
|--LITERAL_STATIC -> static
|--LITERAL_TRANSITIVE -> transitive
|--DOT -> .
`--SEMI -> ;
```

So the Check needs `REQUIRES` in its acceptable tokens and needs to read the modifier children directly rather than through a `MODIFIERS` node.

Expected on the module input:

```
2:14: 'static' modifier out of order, modifiers on 'requires' should appear as 'transitive static'.
```

Correct input, no violations:

```java
module com.example.app {
    requires transitive static java.sql;
    requires transitive java.xml;
    requires static java.compiler;
    requires java.base;
}
```

**Additional context**

This is not purely a token addition, and the reason is worth having in view before implementation.

`ModifierOrder` has enforced the JLS recommendation since 3.0. Its javadoc cites JLS 8.1.1, 8.3.1, 8.4.3 and 9.4, and Google style 4.8.7 defers to the JLS ("appear in the order recommended by the Java Language Specification"), which is why google_checks enables the Check unconfigured.

The `requires` rule has a different source. JLS 7.7.1 writes:

```
RequiresModifier:
(one of)
transitive  static
```

with no ordering statement, and its only constraint is that the same modifier must not appear twice. javac accepts both orders:

```bash
vivek@Viveks-MacBook-Air jlsorder % cat a/module-info.java
module com.example.a {
    requires transitive static java.sql;
}
vivek@Viveks-MacBook-Air jlsorder % cd a && javac module-info.java ; echo "exit $?"
exit 0

vivek@Viveks-MacBook-Air jlsorder % cat b/module-info.java
module com.example.b {
    requires static transitive java.sql;
}
vivek@Viveks-MacBook-Air jlsorder % cd b && javac module-info.java ; echo "exit $?"
exit 0
```

So `transitive static` is a Google convention rather than a JLS recommendation. Two consequences for the implementation:

1. transitive should not be added to the existing JLS_ORDER array. It would work mechanically, but that array is published in the Check documentation as the ordered list of class and member modifiers, and transitive cannot appear on any of them. Its position in the list would also be unjustifiable, since no specification says whether it sorts before or after abstract or private, those modifiers never appearing together. A separate REQUIRES_ORDER = {"transitive", "static"} keeps each list meaning what it says.

2. The existing message reads "out of order with the JLS suggestions", which does not fit here, since the JLS makes no suggestion for `requires` modifiers. A separate message key for the `requires` case keeps both messages accurate about what they are enforcing.

Because the JLS is silent on this order, there is no single correct answer the way there is for class and member modifiers, so a property may be worth having as well, for example `requiresModifierOrder` defaulting to `transitive, static`. That default matches the Google style guide and works out of the box for google_checks, while letting teams that follow a different convention adjust it. The Check has no properties today, so this would be its first, and whether the fixed default is enough is worth deciding during review.

**Comment by romani** [comment]
we need separate Check for this to avoid any conflicts.

do you agree ?

**Comment by vivek-0509** [comment]
I had the same thing in mind, my only hesitation is proportion, like a whole new Check for one rule about the order of two modifiers on requires feels small... but yeah extending modifier check is conflicting, i don't  have better idea we can go with new check...

**Comment by romani** [comment]
if there will be more requests from user on cusomization, being in separate module to cover this will be easy.
in common module, it will challange as need to reconcile config+behavior for all usages. 

Checks can reuse common code through Util classes.

**Comment by vivek-0509** [comment]
Makes sense, agreed... will open issue for new check

**Comment by vivek-0509** [comment]
@romani issue open https://github.com/checkstyle/checkstyle/issues/21168

**Comment by vivek-0509** [comment]
Closed in favour of https://github.com/checkstyle/checkstyle/issues/21168
